File: 606.ConstructStringFromBinaryTree/tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if not t:
            return ''
        if not t.left and not t.right:
            return str(t.val)

        left = self.tree2str(t.left)
        right = self.tree2str(t.right)
        return str(t.val) + '(' + left + ')' + ('(' + right + ')' if right else '')

File: 606.ConstructStringFromBinaryTree/tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    string tree2str(TreeNode* t) {
        if (!t) {
            return "";
        }

        if (!t->left && !t->right) {
            return to_string(t->val);
        }

        string left = tree2str(t->left);
        string right = tree2str(t->right);

        return to_string(t->val) + "(" + left + ")" + (right != "" ? "(" + right + ")" : "");
    }
};
